process/receipts/receiptsRepository.go

```diff
@@ -0,0 +1,115 @@
+package receipts
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go-core/data"
+	"github.com/ElrondNetwork/elrond-go-core/data/batch"
+	"github.com/ElrondNetwork/elrond-go-core/hashing"
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	logger "github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/storage"
+)
+
+var log = logger.GetOrCreate("process/receipts")
+
+type receiptsRepository struct {
+	marshaller        marshal.Marshalizer
+	hasher            hashing.Hasher
+	store             dataRetriever.StorageService
+	emptyReceiptsHash []byte
+}
+
+// NewReceiptsRepository creates a new receiptsRepository
+func NewReceiptsRepository(args ArgsNewReceiptsRepository) (*receiptsRepository, error) {
+	err := args.check()
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", errCannotCreateReceiptsRepository, err)
+	}
+
+	// We cache this special hash (as a struct member)
+	emptyReceiptsHash, err := createEmptyReceiptsHash(args.Marshaller, args.Hasher)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", errCannotCreateReceiptsRepository, err)
+	}
+
+	return &receiptsRepository{
+		marshaller:        args.Marshaller,
+		hasher:            args.Hasher,
+		store:             args.Store,
+		emptyReceiptsHash: emptyReceiptsHash,
+	}, nil
+}
+
+func createEmptyReceiptsHash(marshaller marshal.Marshalizer, hasher hashing.Hasher) ([]byte, error) {
+	emptyReceipts := &batch.Batch{Data: make([][]byte, 0)}
+	return core.CalculateHash(marshaller, hasher, emptyReceipts)
+}
+
+// SaveReceipts saves the receipts in the storer (receipts unit) for a given block header
+func (repository *receiptsRepository) SaveReceipts(holder *process.ReceiptsHolder, header data.HeaderHandler, headerHash []byte) error {
+	storageKey := repository.decideStorageKey(header.GetReceiptsHash(), headerHash)
+
+	log.Debug("receiptsRepository.SaveReceipts()", "headerNonce", header.GetNonce(), "storageKey", storageKey)
+
+	receiptsBytes, err := marshalReceiptsHolder(holder, repository.marshaller)
+	if err != nil {
+		return err
+	}
+
+	// Nothing to be saved
+	if len(receiptsBytes) == 0 {
+		return nil
+	}
+
+	err = repository.getStorer().Put(storageKey, receiptsBytes)
+	if err != nil {
+		return fmt.Errorf("%w: %v", errCannotSaveReceipts, err)
+	}
+
+	return nil
+}
+
+func (repository *receiptsRepository) getStorer() storage.Storer {
```

Comment:
instead of using this function, why not storing the receipts storer instead of the storer service?

Reply:
Refactored.